src/main/groovy/com/ullink/gradle/sonar/SonarScannerMsbuildPlugin.groovy

```diff
@@ -0,0 +1,94 @@
+package com.ullink.gradle.sonar
+
+import de.undercouch.gradle.tasks.download.Download
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.tasks.Copy
+import org.gradle.api.tasks.Exec
+
+/**
+ * Uses SonarScanner for MSBuild to run the Sonarqube analysis.
+ * See https://docs.sonarqube.org/display/SCAN/Analyzing+with+SonarQube+Scanner+for+MSBuild
+ * Sonar properties can be configured using the same syntax as for the org.sonarqube plugin
+ * in order ro keep compatibility.
+ */
+class SonarScannerMsbuildPlugin implements Plugin<Project> {
+
+    static final DEFAULT_SCANNER_VERSION = '4.6.0.1930'
+    static final DEFAULT_MSBUILD_TASK_NAME = 'msbuild'
+    static final File CACHE_DIR = new File(System.getProperty('user.home'), '.gradle/caches/sonar-scanner-msbuild')
+    static final PROJECT_KEY_ARG = 'sonar.projectKey'
+    static final MANDATORY_ARGS = [n: 'sonar.projectName', v: 'sonar.projectVersion']
+
+    @Override
+    void apply(Project project) {
+        def scannerExtension = project.extensions.create('sonarqube', SonarScannerMsbuildExtension)
+        project.afterEvaluate {
```

Comment:
I know I've written a few myself (way too many :-( )
But I think those blocks are somewhat evil

Wouldn't it be possible to just have tasks do the work?

Reply:
I partially agree. The execution order of the `afterEvaluate` blocks is not guaranteed and having many of them could make the build unpredictable.
Unfortunately in order to make msbuild depend on the sonar tasks (see the discussion above) I must make sure that the msbuild task is evaluated (created).